fix(internal): write note orphans child notes

hikchoi opened the pull request:
# fix: write note orphans child notes

This PR:
- fixes a bug where overwriting an existing note's id will result in:
  - stale children props of the parent of the note
  - orphaned children of the written note
- generally refactors code that this edge case covers in `_writeNewNote` and add omitted relevant note change entries.
- adds the test that was skipped in `EngineEvents.spec.ts` that this bug was blocking.

No added circ dependencies.

## Code

### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones 
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended

- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed 
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)



## Instrumentation

### Basics

- [~] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Extended

- [~] can we track the performance of this change to know if it is _successful_? 
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

## 



## Tests

### Basics

- [x] [Write Tests](dev.process.qa.test) 
- [x] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [~] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended

- [~] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](dev.ref.test-workspace)

- CSS
  - [~] display is correct for following dimensions
    - [~] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [~] lg: screen ≥ 992px
    - [~] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [~] display is correct for following browsers (across the various dimensions)
    - [~] safari
    - [~] firefox
    - [~] chrome



## Docs

### Basics

- [~] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

## 


### Basics

- [~] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

## 



## Close the Loop

### Basics

### Extended

- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)
  - eg. [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

hikchoi commented:
@Harshita-mindfire This PR touches on some of the `GithubIssuePod.spec.ts` tests that were breaking from the changes. Can you check if these changes are fine? 🙏🏼

jonathanyeung commented on packages/common-all/src/dnode.ts:
```diff
@@ -346,6 +346,35 @@ export class NoteUtils {
+  }): NoteChangeEntry[] {
+    const changed: NoteChangeEntry[] = [];
+    const { childToDelete, notes } = opts;
+    // const parent = DNodeUtils.getParent(childToDelete, { nodeDict: notes });
```
nit: Delete commented line.

jonathanyeung commented on packages/engine-test-utils/src/__tests__/engine-server/EngineEvents.spec.ts:
```diff
@@ -155,19 +153,52 @@ describe("GIVEN a DendronEngineClient running on client-side", () => {
+                expect(deleteEntries.length).toEqual(1);
+
+                expect(createEntries[0].note.fname).toEqual("foo");
+                expect(createEntries[0].note.id).toEqual("updatedID");
```
Can we add 2 more checks on the foo note? 
- That its parent is correct (root I think here)
- That its children are correct (foo.ch1)

Although it does look like you added these checks to the other Engine test, so maybe we're already covered.

jonathanyeung commented on packages/engine-test-utils/src/__tests__/engine-server/EngineEvents.spec.ts:
```diff
@@ -155,19 +153,52 @@ describe("GIVEN a DendronEngineClient running on client-side", () => {
           );
 
-          await engineClient.writeNote(fooUpdated, { updateExisting: true });
+          await engineClient.writeNote(fooUpdated);
```
It's still not very clear to me on whether the `{updateExisting: true}` flag is necessary to be set here or not to trigger this behavior. 

IMO - this _should_ be the correct behavior regardless of whether this flag is set or not - @kevinslin thoughts?

hikchoi replied:
I don't think for this case setting `updateExisting: true` is right. we want to overwrite.

in `engine.writeNote` if `updateExisting: true` is set, we don't go into the logic at all where the overwrite happens.

hikchoi replied:
for anything other than changing the note's id, `updateExisting: true` makes sense.

jonathanyeung commented:
Looking good. 

There's a potentially related failure in the plugin CI pass for `MoveNoteCommand`.

hikchoi commented:
> There's a potentially related failure in the plugin CI pass for MoveNoteCommand.

Looking into it now. will ping when it's fixed.

hikchoi commented:
@jonathanyeung 

update: 

- cleaned up comments
- added mentioned cases to test
- cleaned up `_writeNewNote` logic so that it's updating _existing note`s parent_ instead of _new note's parent_. They will be the same since we are overwriting them in place, but semantically this will be more natural.
- move note command's test was failing because the deprecated `createNote` was not ensuring the testing note's existence in engine. swapped it out to the new `createNoteWithEngine` so the test works as intended.

hikchoi commented:
also updated the fixed tests in `GithubIssuePod.spec.ts` to not use non-null assertions as per our recent effort to remove all type assertions and casting.

Harshita-mindfire reviewed (approved):
lgtm!
